### PR TITLE
feat: add a widget level loading indicator for x-y plot #2379

### PR DESCRIPTION
**Pull request by Priyanshu44**
Issue: https://github.com/awslabs/iot-app-kit/issues/2379

Verification:
-> Added widget level loading indicator on refresh if it continues for 3 sec or more.
-> Widget level indicator should only show when the "primary" loading indicator is not being shown (the one displayed in the 
     center of the "chart area" )
-> Changed "isRefreshing" logic in core package and made respective changes.

[secondary-loading-indicator.webm](https://github.com/awslabs/iot-app-kit/assets/81667589/9d673a35-a58f-4ce4-970d-eb626a458196)

### CODE REVIEW

File: packages/dashboard/src/components/widgets/tile/tile.tsx [hunk 1]
```diff
@@ -87,6 +90,24 @@ const WidgetTile: React.FC<WidgetTileProps> = ({ children, widget, title, remove
     });
   };
 
+  const [delayLoading, setDelayLoading] = useState(isRefreshing);
+  const timeout = useRef<number | null | NodeJS.Timeout>(null);
+  const delay = 3000;
```
Comment (mnischay) [inline packages/dashboard/src/components/widgets/tile/tile.tsx:95]: please  move this to a constants file

File: packages/dashboard/src/customization/widgets/lineScatterChart/component.tsx [hunk 2]
```diff
@@ -25,6 +25,8 @@ import NoChartData from '../components/no-chart-data';
 import { default as lineSvgDark } from './line-dark.svg';
 import { IoTSiteWiseDataStreamQuery } from '~/types';
 import { assetModelQueryToSiteWiseAssetQuery } from '../utils/assetModelQueryToAssetQuery';
+// eslint-disable-next-line no-restricted-imports
+import { useVisualizedDataStreams } from '@iot-app-kit/react-components/src/components/chart/hooks/useVisualizedDataStreams';
```
Comment (mnischay) [inline packages/dashboard/src/customization/widgets/lineScatterChart/component.tsx:29]: why the eslint exceptions ? you should not be importing from src, if you need this exposed, updated the react-components public exports file

**Review by mnischay — CHANGES_REQUESTED** [review]

File: packages/dashboard/src/customization/widgets/lineScatterChart/component.tsx (hunk 2 above)
Comment (Priyanshu44) [reply packages/dashboard/src/customization/widgets/lineScatterChart/component.tsx]: Resolved

File: packages/dashboard/src/components/widgets/tile/tile.tsx (hunk 1 above)
Comment (Priyanshu44) [reply packages/dashboard/src/components/widgets/tile/tile.tsx]: resolved

File: packages/dashboard/src/components/widgets/tile/contants.ts [hunk 3]
```diff
@@ -0,0 +1 @@
+export const delay = 3000;
```
Comment (mnischay) [inline packages/dashboard/src/components/widgets/tile/contants.ts:1]: const are ALL CAPS and should denote the units 

`USE_CASE_DELAY_IN_SECS
`

File: packages/dashboard/src/components/widgets/tile/tile.spec.tsx [hunk 4]
```diff
@@ -98,4 +98,42 @@ describe('WidgetTile', () => {
 
     expect(container.querySelector('[aria-label="delete widget"]')).not.toBeInTheDocument();
   });
+
+  it('should render secondary loading indicator on isRefreshing', function () {
```
Comment (mnischay) [inline packages/dashboard/src/components/widgets/tile/tile.spec.tsx:102]: may be you can test the 3 sec logic ?

File: packages/dashboard/src/components/widgets/tile/contants.ts (hunk 3 above)
Comment (Priyanshu44) [reply packages/dashboard/src/components/widgets/tile/contants.ts]: resolved

File: packages/dashboard/src/components/widgets/tile/tile.spec.tsx (hunk 4 above)
Comment (Priyanshu44) [reply packages/dashboard/src/components/widgets/tile/tile.spec.tsx]: resolved

File: packages/dashboard/src/components/widgets/tile/tile.tsx [hunk 5]
```diff
@@ -87,6 +91,23 @@ const WidgetTile: React.FC<WidgetTileProps> = ({ children, widget, title, remove
...
+        setDelayLoading(isRefreshing);
+      }, USE_CASE_DELAY_IN_SECS);
+    } else {
+      setDelayLoading(false);
+    }
+    return () => {
+      clearTimeout(timeout.current as number);
+    };
+  }, [isRefreshing]);
+
```
Comment (ssjagad) [inline packages/dashboard/src/components/widgets/tile/tile.tsx:110]: This clutters the code base a bit and is re-usable. Let's pull this out into a custom hook

File: packages/dashboard/src/components/widgets/tile/contants.ts [hunk 6]
```diff
@@ -0,0 +1 @@
+export const USE_CASE_DELAY_IN_SECS = 3000;
```
Comment (ssjagad) [inline packages/dashboard/src/components/widgets/tile/contants.ts:1]: Constant name is confusing. How about something like DATA_DELAY_AFTER _REQUEST_IN_SECS

File: packages/core/src/data-module/data-cache/dataReducer.ts [hunk 7]
```diff
@@ -171,8 +176,9 @@ export const dataReducer: Reducer<DataStreamsStore, AsyncActions> = (
           : requestCache,
         dataCache: updatedDataCache,
         isLoading: false,
-        isRefreshing: !!restOfDataStream.isRefreshing,
         error: undefined,
+        numOutgoingRequests: noOfOutGoingrequest,
+        isRefreshing: noOfOutGoingrequest ? noOfOutGoingrequest > 0 : false,
```
Comment (Chandru-HM) [inline packages/core/src/data-module/data-cache/dataReducer.ts:181]: can we use like this " isRefreshing: noOfOutGoingrequest > 0,"  ?

File: packages/dashboard/src/components/widgets/tile/contants.ts [hunk 8]
```diff
@@ -0,0 +1 @@
+export const IS_REFRESHING_DELAY_IN_SECS = 3000;
```
Comment (Chandru-HM) [inline packages/dashboard/src/components/widgets/tile/contants.ts:1]: can we use like ""ISREFRESHING_DELAY"

**Review by Chandru-HM — CHANGES_REQUESTED** [review]

File: packages/core/src/data-module/data-cache/dataReducer.ts [hunk 9]
```diff
@@ -67,7 +70,8 @@ export const dataReducer: Reducer<DataStreamsStore, AsyncActions> = (
           : requestCache,
         id,
         isLoading,
-        isRefreshing: !!streamStore?.isRefreshing,
+        isRefreshing: true,
+        numOutgoingRequests: noOfOutGoingrequest,
```
Comment (Chandru-HM) [inline packages/core/src/data-module/data-cache/dataReducer.ts:74]: can we directly use "++outGoingrequest" here? why need extra constant.

Comment (Priyanshu44) [reply packages/core/src/data-module/data-cache/dataReducer.ts]: resolved

File: packages/core/src/data-module/data-cache/dataReducer.ts (hunk 7 above)
Comment (Priyanshu44) [reply packages/core/src/data-module/data-cache/dataReducer.ts]: resolved

File: packages/dashboard/src/components/widgets/tile/tile.tsx (hunk 5 above)
Comment (Priyanshu44) [reply packages/dashboard/src/components/widgets/tile/tile.tsx:115]: resolved

**Comment by Priyanshu44** [comment]
created another CR so closing this one.

**Pull request closed without merging**